crates/ursa-pod/src/client.rs

```diff
@@ -0,0 +1,158 @@
+            Poll::Ready(_) => todo!("handle errors"),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let size = self.content_len as usize;
+        (size, Some(size))
+    }
+}
+
+/// UFDP Client. Accepts any stream of bytes supporting [`AsyncRead`] + [`AsyncWrite`]
+pub struct UfdpClient<S: AsyncRead + AsyncWrite + Unpin> {
+    pub transport: Framed<S, UrsaCodec>,
+}
+
+impl<S> UfdpClient<S>
+where
+    S: AsyncRead + AsyncWrite + Unpin + Send,
+{
+    /// Create a new client, attempting to handshake with the destination
+    pub async fn new(stream: S) -> Result<Self, UrsaCodecError> {
+        let mut transport = Framed::new(stream, UrsaCodec::default());
+
+        // send handshake
+        transport
+            .send(UrsaFrame::HandshakeRequest {
+                version: 0,
+                supported_compression_bitmap: 0,
+                lane: 0xFF,
+                pubkey: [1; 48],
+            })
+            .await
+            .expect("handshake request");
+
+        // receive handshake
+        if let Ok(frame) = transport.next().await.expect("handshake response") {
+            match frame {
+                UrsaFrame::HandshakeResponse { .. } => {
+                    info!("received handshake response from server: {frame:?}");
+                }
+                f => return Err(UrsaCodecError::UnexpectedFrame(f.tag().unwrap())),
+            }
+        }
+
+        Ok(Self { transport })
+    }
+
+    /// Send a request for content.
+    pub async fn request(&mut self, hash: Blake3Cid) -> Result<UfdpResponse<S>, UrsaCodecError> {
+        self.transport
+            .send(UrsaFrame::ContentRequest { hash })
+            .await
+            .expect("content request");
+
+        match self.transport.next().await.expect("content response")? {
+            UrsaFrame::ContentResponse {
+                content_len,
+                proof_len,
+                ..
+            } => {
+                debug!("received content response, streaming proof ({proof_len})");
+                if proof_len != 0 {
+                    unimplemented!()
+                    // todo:
+                    //  - stream and collect blake3tree bytes
+                    //  - decode it
+                    //  - pass to UfdpResponse to incrementally verify content
+                }
+
+                if content_len != 0 {
+                    self.transport
+                        .codec_mut()
+                        .read_buffer(content_len as usize, CONTENT_CHUNK_LEN);
+
+                    Ok(UfdpResponse {
+                        client: self,
+                        // todo: proof
+                        current_block: BytesMut::with_capacity(256 * 1024),
+                        index: 0,
+                        content_len,
+                    })
+                } else {
+                    Err(UrsaCodecError::Unknown)
+                }
```

Comment:
move this logic to `UfdpResponse::new().await?`, but could be done later too, because we want the UfdpResponse to handle the proof.